llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp

```diff
@@ -1159,8 +1159,12 @@ void VectorLegalizer::Expand(SDNode *Node, SmallVectorImpl<SDValue> &Results) {
   }
 
   SDValue Unrolled = DAG.UnrollVectorOp(Node);
-  for (unsigned I = 0, E = Unrolled->getNumValues(); I != E; ++I)
-    Results.push_back(Unrolled.getValue(I));
+  const LoadSDNode *Ld = dyn_cast<LoadSDNode>(Unrolled.getNode());
+  if (Ld)
+    Results.push_back(Unrolled);
+  else
+    for (unsigned I = 0, E = Unrolled->getNumValues(); I != E; ++I)
+      Results.push_back(Unrolled.getValue(I));
```

Comment:
Special casing the return values is the wrong fix. UnrollVectorOp should have a consistent return scheme. The accounting in UnrollVectorOp isn't accounting for chain operands; I think the getNumValues == 2 check in UnrollVectorOp should be discounting chain outputs

Reply:
The SDag looks like this
  t39: v2i32,ch = load<(dereferenceable invariant load (s64) from %ir..kernarg.offset, align 4, addrspace 4)> t0, t11, undef:i64
              t37: i32 = extract_vector_elt t39, Constant:i32<0>
            t38: i16 = truncate t37
              t35: i32 = extract_vector_elt t39, Constant:i32<1>
            t36: i16 = truncate t35
          t20: v2i16 = BUILD_VECTOR t38, t36
        t32: v2i32 = any_extend t20
        
It looks like the logic after 'assert(N->getNumValues() == 1 ' is responsible for returning load.

Reply:
Scalars var gets both extract_vectors and get_build_vector coalesces them and ends up just returning t39 load

Reply:
But what is being unrolled? UnrollVectorOp doesn't handle loads

Reply:
the any_extend (t32) is being unrolled. The last line in UnrollVectorOps calls getBuildVector with the two extract_vector_elt expressions (t35, t37). Since those extract_vector_elt reference the same vector (t39) and extract consecutive indices, the FoldBUILD_VECTOR function returns the load (t39). So, the result from UnrollVectorOps contains two values.

Reply:
Is the correct fix just to use Node->getNumValues() instead of Unrolled->getNumValues()?

Reply:
Well even that's not technically right. If there is a single result expected, we shouldn't call getValue we should use the SDValue directly as the result. Folding might have allowed the SDValue to point to one result of a node with multiple results where the desired result isn't result 0.

If multiple results are expected, it should be a MERGE_VALUES node which isn't subject to folding.

Something like this is probably closer to correct.

```
if (Node->getValues() == 1) {
  // Only 1 result expect, use it directly. It isn't guaranteed to be result 0.
  Results.push_back(Unrolled);
} else {
  assert(Node->getValues() > 1);
  assert(Unrolled->getOpcode() == ISD::MERGE_VALUES);
  for (unsigned I = 0, E = Node->getNumValues(); I != E; ++I)
    Results.push_back(Unrolled.getValue(I));
}
```

Reply:
Thanks for the suggestion. I think it makes sense to have it work similar to LowerOperationWrapper.

Reply:
Thank you all for the feedback. I took @topperc 's suggestion and replicated the lowering logic in the newest update.